### ISSUE TITLE
sql/opt: crash while planning `//` binop over two CASE expressions

### ISSUE THREAD

**Issue by knz** [cockroachdb/cockroach#35177]
The following SQL crashes the CBO:

```sql
VALUES (
        -3993911422377595656
        << array_position(ARRAY[-0.48442661814333066, 0.08491291889752639, -0.017100705152083395, -1.5940068386193618], NULL)
       ),
       (-805133545430125191),
       (
        4085601939627203299
        - CASE
            WHEN false THEN CASE WHEN NULL THEN 3326209048762244771 ELSE -254531168219194454 END
            ELSE NULL
            END
            // CASE WHEN NULL THEN 6010968360109510481 ELSE NULL END
       );
```

Stack trace:

```
        panic: could not find overload for binary expression floor-div [recovered]


github.com/cockroachdb/cockroach/pkg/sql.(*connExecutor).closeWrapper(0xc001664000, 0x383dc40, 0xc00164f140, 0x2ba9c80, 0xc000476690)
        /home/spill/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:721 +0x36d
github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1(0xc001664000, 0x383dc40, 0xc00164f140)
        /home/spill/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:427 +0x61
panic(0x2ba9c80, 0xc000476690)
        /usr/lib/go-1.11/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).Build.func1(0xc0014ba990)
        /home/spill/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/builder.go:142 +0x9a
panic(0x2ba9c80, 0xc000476690)
        /usr/lib/go-1.11/src/runtime/panic.go:513 +0x1b9
github.com/cockroachdb/cockroach/pkg/sql/opt/memo.BinaryAllowsNullArgs(0xc001650035, 0x385b120, 0x550eec0, 0x385b120, 0x550eec0, 0x43fb247b)
        /home/spill/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/memo/typing.go:80 +0x132
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*CustomFuncs).AllowNullArgs(0xc001664ba8, 0x5500035, 0x385f740, 0xc001657ba0, 0x385f740, 0xc001657ba0, 0x2e06080)
        /home/spill/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/custom_funcs.go:1090 +0x93
github.com/cockroachdb/cockroach/pkg/sql/opt/norm.(*Factory).ConstructFloorDiv(0xc001664b98, 0x385f740, 0xc001657ba0, 0x385f740, 0xc001657ba0, 0xc0014b9d01, 0x0)
        /home/spill/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/norm/factory.og.go:12298 +0x4d6
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).constructBinary(0xc001222690, 0x7, 0x385f740, 0xc001657ba0, 0x385f740, 0xc001657ba0, 0x385ae20, 0x550eec0,
0xc001657ba0, 0xc0012b7080)
        /home/spill/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:635 +0x383
github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder.(*Builder).buildScalar(0xc001222690, 0x3858180, 0xc00173d340, 0xc0000f2800, 0x0, 0x0, 0x0, 0x385e540, 0xc0012b7080)
        /home/spill/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/optbuilder/scalar.go:195 +0xc7a
```

This doesn't repro with `set optimizer = off`

**Comment by knz** [comment]
Note: as discussed previously, I kindly request that you implement a catch-all to translate all unhandled panics in the CBO to `pgerror.NewAssertionErrorf` at the top level, so we avoid this type of crash in the 19.1 release. Thanks

**Comment by jordanlewis** [comment]
Looks like a duplicate of #34930.

**Comment by rytaft** [comment]
This was fixed by #35188.
